## Task 1: Model Validation

**part a:**  
- We found our best-performing model to be the **Random Forest Classifier** model, which had the following results: 

In [12]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix, ConfusionMatrixDisplay
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression
from sklearn.svm import LinearSVC
from sklearn.model_selection import train_test_split


In [13]:
df = pd.read_csv("newdata (2).csv")
X = df[['popularity', 'danceability', 'valence', 'tempo']]
y = df['winner']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [14]:
#Create the Random Forest Classifier model
rf_model = RandomForestClassifier(n_estimators=250, class_weight='balanced', random_state=42)

#Train the  Random Forest Classifier model on the data
rf_model.fit(X_train, y_train)

#Utilize the trained model to make predictions on the set
y_pred_rf = rf_model.predict(X_test)

#Calculate the performance metrics             
acc_rf = accuracy_score(y_test, y_pred_rf)
prec_rf = precision_score(y_test, y_pred_rf, zero_division=0)
rec_rf = recall_score(y_test, y_pred_rf, zero_division=0)

print("Random Forest Classifier Results: ")
print(f"Accuracy:  {acc_rf:.4f}")
print(f"Precision: {prec_rf:.4f}")
print(f"Recall:    {rec_rf:.4f}")

Random Forest Classifier Results: 
Accuracy:  0.9904
Precision: 1.0000
Recall:    0.6626


- We'll now perform cross validation using the k-folds method to return our model's stability

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split, cross_val_score, StratifiedKFold
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score

df = pd.read_csv("newdata (2).csv")
X = df[['popularity', 'danceability', 'valence', 'tempo']]
y = df['winner']

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Best model (can tune later if needed)
rf_model = RandomForestClassifier(n_estimators=250, class_weight='balanced', random_state=42)

# Train on the training set
rf_model.fit(X_train, y_train)

# Predict
y_pred_rf = rf_model.predict(X_test)
y_pred_proba_rf = rf_model.predict_proba(X_test)[:, 1]

# Normal evaluation
acc_rf = accuracy_score(y_test, y_pred_rf)
prec_rf = precision_score(y_test, y_pred_rf, zero_division=0)
rec_rf = recall_score(y_test, y_pred_rf, zero_division=0)
f1_rf = f1_score(y_test, y_pred_rf)
auc_rf = roc_auc_score(y_test, y_pred_proba_rf)

print("Random Forest Classifier Test Results:")
print(f"Accuracy:  {acc_rf:.4f}")
print(f"Precision: {prec_rf:.4f}")
print(f"Recall:    {rec_rf:.4f}")
print(f"F1 Score:  {f1_rf:.4f}")
print(f"AUC:       {auc_rf:.4f}")

# ---- CROSS VALIDATION ----
cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

acc_scores = cross_val_score(rf_model, X, y, cv=cv, scoring='accuracy')
f1_scores = cross_val_score(rf_model, X, y, cv=cv, scoring='f1')
auc_scores = cross_val_score(rf_model, X, y, cv=cv, scoring='roc_auc')

print("\nCross-Validation Results (Random Forest):")
print(f"Accuracy: Mean={acc_scores.mean():.4f}, Std={acc_scores.std():.4f}")
print(f"F1 Score: Mean={f1_scores.mean():.4f}, Std={f1_scores.std():.4f}")
print(f"AUC:      Mean={auc_scores.mean():.4f}, Std={auc_scores.std():.4f}")


Random Forest Classifier Test Results:
Accuracy:  0.9904
Precision: 1.0000
Recall:    0.6626
F1 Score:  0.7970
AUC:       0.8852


**part b:**  
- With an F1 score of .7920, which is a significantly high score (> .7), we know that our model is able to **consistently make precise predictions on our dataset**.
- Our AUC also returns an objectively high (> .7) score of .8852, reflecting its behavior across random thresholds and its ability to distinguish between 